clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -3630,6 +3630,14 @@ LangAS CodeGenModule::GetGlobalVarAddressSpace(const VarDecl *D) {
     return AddrSpace;
   }
 
+  if (LangOpts.SYCLIsDevice) {
+    if (!getenv("DISABLE_INFER_AS")) {
+      if (!D || D->getType().getAddressSpace() == LangAS::Default) {
```

Comment:
Shouldn't we use constant address space if variable has const qualifier?

Reply:
constant address space != const qualifier.

The major difference is:
- constant pointer points to a memory that cannot be changed during a kernel invocation
- const pointer points to a memory that cannot be changed *through that pointer*, but can be changed through a non-const pointer

```
void fun() {
  static const int foo = 42;
  int bar = 43;
  notfun(fooOrBar? &foo : &bar);
  bar = 44;
  notfun(fooOrBar? &foo : &bar);
}

void notfun(const int *p) {
}
```

The code above is illegal if a static const variable resides in constant address space.